Condensing diffusion instead of transport

mkreher13 opened the pull request:
I wanted to build on our discussion from the weekend. What was recently merged was a fix that took a completely broken diffusion coefficient condensing function and made it workable by condensing all the underlying tallies which amounts to condensing the transport cross section and then re-calculating the diffusion coefficient. In this case, condensed_diffusion = 1/(3*condensed_tr). 

However, per our conversation, it is more accurate to condense the diffusion coefficient instead (which amounts to condensing 1/tr). In this PR, I believe I've done that. As you can see in the results of mgxs_library_condense, condensed_diffusion is NOT equal to 1/(3*condensed_tr). 

I think a knowledgeable user will score the diffusion coefficient on a very fine energy grid, and then use the condense function to get the diffusion coefficient on the energy grid they actually want to use for the problem. 

The changes in xs_tally are necessary because when condensing, the diffusion coefficient is calculated, so you can't re-calculate it in xs_tally later (for instance, you can't re-slice the P1 scattering tally since it no longer exists after slicing it once). This change ensures that get_condensed_xs and xs_tally are compatible. 

I'm hoping for feedback. Did I make any mistakes? What does everybody think? @gridley @nelsonag

nelsonag commented on openmc/mgxs/mgxs.py:
```diff
@@ -3154,30 +3154,133 @@ def xs_tally(self):
+        condensed_xs = copy.deepcopy(self)
+
+        if condense_diff_coef:
             p1_tally = self.tallies['scatter-1']
```
Would this not fail if the user tried condensing an MGXS object already condensed? (back to the above: `self.tallies['scatter-1']` wouldnt be present.

nelsonag commented on tests/regression_tests/mgxs_library_condense/results_true.dat:
```diff
@@ -214,16 +214,16 @@
-2      1  2  1        1   total  0.779112  0.047034
-1      2  1  1        1   total  0.787475  0.050368
-3      2  2  1        1   total  0.769656  0.058065
+0      1  1  1        1   total  0.931945  0.045510
```
I presume you verified these new values with hand-calculations of the pre-condensed results?

nelsonag reviewed (changes requested):
Hi @mkreher13! Sorry for the delay on this one. I have attached some comments which should be addressed. But at a bigger picture: I dont quite like the idea of `self.tallies` being different depending on if it is a previously-condensed MGXS or not (if I'm understanding correctly).

Is there a clean way to create a cloned MGXS object, compute xs_tally on that, and then condense the clone?  That way the original is 'untouched'. Thanks!

mkreher13 commented:
Sorry for waiting 9 months to address your comments, as I have had other priorities. I am circling back to it as I am wrapping up my PhD, and I need good diffusion coefficients for my thesis. This piece of code is actually pretty important, as it is crucial to be able to finely tally the transport cross section and then condense the diffusion coefficients. 

I addressed your concerns by including a comment that described the role of the if-statement relating to `'scatter-1'.` If `'scatter-1'` is not in `self.tallies`, it is not because the MGXS has been condensed, but because the transport correction has already been applied. If that is the case, it is safe to jump ahead to the main body of the calculation. Otherwise, the transport correction needs to be done first. 

You were correct that if `'scatter-1'` is not present, you couldn't re-condense a condensed diffusion coefficient. Essentially, you can always compute multiple condensations directly from the original MGXS by using:
`mgxs_lib.get_mgxs(cell, 'diffusion-coefficient').get_condensed_xs(coarse_groups)` several times in a row with coarser and coarser groups.
But you couldn't do:
`condensed_dc = mgxs_lib.get_mgxs(cell, 'diffusion-coefficient').get_condensed_xs(coarse_groups)`
`condensed_dc .get_condensed_xs(coarser_groups)`

To fix this in `get_condensed_xs()`, I added the if-statement relating to `'scatter-1' `which again ensures that the transport correction is done before jumping into the main calculation. Now there is no issue stringing multiple `get_condensed_xs()` calls one after another. 

I also removed the variable `condense_diff_coef`. That variable existed to give the user the option to condense the underlying tallies rather than the diffusion coefficients themselves. My colleague pointed out that if that option is not "good physics" then it shouldn't be an option. So by removing it, the user has to use the better option of condensing the diffusion coefficients instead. 

I also did a few hand calculations to make sure that the condensation was correct.

nelsonag commented:
Thanks @mkreher13 and apologies for the late review! Good luck with the remainder of your dissertation!
